src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,237 @@
+
+namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
+{
+    internal class InlineDiagnosticsAdornmentManager : AdornmentManager<InlineDiagnosticsTag>
+    {
+        private readonly IClassificationTypeRegistryService _classificationRegistryService;
+        private readonly IClassificationFormatMap _formatMap;
+
+        public InlineDiagnosticsAdornmentManager(
+            IThreadingContext threadingContext, IWpfTextView textView, IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IAsynchronousOperationListener asyncListener, string adornmentLayerName,
+            IClassificationFormatMapService classificationFormatMapService,
+            IClassificationTypeRegistryService classificationTypeRegistryService)
+            : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
+        {
+            _classificationRegistryService = classificationTypeRegistryService;
+            _formatMap = classificationFormatMapService.GetClassificationFormatMap(textView);
+            _formatMap.ClassificationFormatMappingChanged += OnClassificationFormatMappingChanged;
+            TextView.ViewportWidthChanged += TextView_ViewportWidthChanged;
+        }
+
+        private void TextView_ViewportWidthChanged(object sender, EventArgs e)
+        {
+            if (AdornmentLayer is not null)
+            {
+                var elements = AdornmentLayer.Elements;
+                if (elements.Any())
+                {
+                    var location = ((InlineDiagnosticsTag)elements[0].Tag).Location;
+                    if (location == InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                    {
+                        var normalizedCollectionSpan = new NormalizedSnapshotSpanCollection(TextView.TextViewLines.FormattedSpan);
+                        UpdateSpans_CallOnlyOnUIThread(normalizedCollectionSpan, true);
+                    }
+                }
+            }
+        }
+
+        private void OnClassificationFormatMappingChanged(object sender, EventArgs e)
+        {
+            if (AdornmentLayer is not null)
+            {
+                var elements = AdornmentLayer.Elements;
+                foreach (var element in elements)
+                {
+                    var tag = (InlineDiagnosticsTag)element.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var format = GetFormat(classificationType);
+                    tag.UpdateColor(format, element.Adornment);
+                }
+            }
+        }
+
+        private TextFormattingRunProperties GetFormat(IClassificationType classificationType)
+        {
+            return _formatMap.GetTextProperties(classificationType);
+        }
+
+        /// <summary>
+        /// Get the spans located on each line so that it can only display the first one that appears on the line
```

Comment:
not really sure i understand the comment.

Reply:
also, if we only display the first one... why do we return a List<...> ?

Reply:
So I was keeping track of a list because, at that point, I still was not sure how we would be handling multiple errors. I removed them now